ui/selectors/assets.ts

```diff
@@ -40,3 +59,152 @@ export function getAssetsMetadata(state: AssetsState) {
+    currentNetwork,
+  ) => {
+    const { hideZeroBalanceTokens } = preferences;
+    const selectedAccountTokenBalancesAcrossChains =
+      tokenBalances[selectedAccountAddress];
+
+    // we need to filter Testnets
+    const isTestNetwork = TEST_CHAINS.includes(currentNetwork.chainId);
+    const filteredAccountTokensChains = Object.fromEntries(
+      Object.entries(selectedAccountTokensChains).filter(([chainId]) =>
+        isTestNetwork
+          ? TEST_CHAINS.includes(chainId as (typeof TEST_CHAINS)[number])
+          : !TEST_CHAINS.includes(chainId as (typeof TEST_CHAINS)[number]),
+      ),
+    );
+    const tokensWithBalance: TokenWithFiatAmount[] = [];
+    Object.entries(filteredAccountTokensChains).forEach(
+      ([stringChainKey, tokens]) => {
+        const chainId = stringChainKey as Hex;
+        const tokenList = tokens as Token[];
+        tokenList.forEach((token: Token) => {
+          const { isNative, address, decimals } = token;
+          const balance =
+            calculateTokenBalance({
+              isNative,
+              chainId,
+              address,
+              decimals,
+              nativeBalances,
+              selectedAccountTokenBalancesAcrossChains,
+            }) || '0';
+
+          const tokenFiatAmount = calculateTokenFiatAmount({
+            token,
+            chainId,
+            balance,
+            marketData,
+            currencyRates,
+          });
+
+          // Respect the "hide zero balance" setting (when true):
+          // - Native tokens should always display with zero balance when on the current network filter.
+          // - Native tokens should not display with zero balance when on all networks filter
+          // - ERC20 tokens with zero balances should respect the setting on both the current and all networks.
+
+          // Respect the "hide zero balance" setting (when false):
+          // - Native tokens should always display with zero balance when on the current network filter.
+          // - Native tokens should always display with zero balance when on all networks filter
+          // - ERC20 tokens always display with zero balance on both the current and all networks filter.
+          if (
+            !hideZeroBalanceTokens ||
+            balance !== '0' ||
+            (token.isNative && isOnCurrentNetwork)
+          ) {
+            tokensWithBalance.push({
+              ...token,
+              balance,
+              tokenFiatAmount,
+              chainId,
+              string: String(balance),
+              primary: '',
+              secondary: 0,
+              title: '',
+            });
+          }
+        });
+      },
+    );
+    return tokensWithBalance;
+  },
+);
+
+export const getMultiChainAssets = createDeepEqualSelector(
+  (_state, selectedAccount) => selectedAccount,
+  getMultichainBalances,
+  getAccountAssets,
+  getAssetsMetadata,
+  getAssetsRates,
+  (_state) => _state.metamask.conversionRates,
```

Comment:
This is the same ?

Reply:
Yes i think there was a refactor commit that introduced this bug; thnx for catching this, updated here [7cea61b](https://github.com/MetaMask/metamask-extension/pull/30389/commits/7cea61b7b028720cef3c9419bc25e7c98aab6080)